libyul/optimiser/MemoryStoreRemover.cpp

```diff
@@ -0,0 +1,311 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/MemoryStoreRemover.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/SSAValueTracker.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/backends/wasm/WasmDialect.h>
+
+#include <libyul/Exceptions.h>
+
+#include <libevmasm/Instruction.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsmtutil/SolverInterface.h>
+
+#include <libsolutil/cxx20.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+using namespace solidity::util;
+using namespace solidity::evmasm;
+using namespace solidity::smtutil;
+
+namespace
+{
+
+/// A helper class that would replace `mstore(key, expression)` for all `key` in
+/// `m_keys`, by `pop(expression)`.
+class EraseMemoryStores: public ASTModifier
+{
+public:
+	EraseMemoryStores(Block& _ast, set<YulString> const& _keys, Dialect const& _dialect):
+		m_keys(_keys),
+		m_dialect(_dialect)
+	{
+		(*this)(_ast);
+	}
+
+private:
+	using ASTModifier::operator();
+
+	void operator()(FunctionCall& _functionCall) override;
+
+	set<YulString> const& m_keys;
+	Dialect const& m_dialect;
+};
+
+void EraseMemoryStores::operator()(FunctionCall& _functionCall)
+{
+	ASTModifier::operator()(_functionCall);
+
+	if (_functionCall.functionName.name == m_dialect.memoryStoreFunction({})->name)
+		if (Identifier const* identifier = get_if<Identifier>(&_functionCall.arguments.at(0)))
+			if (m_keys.count(identifier->name))
+			{
+				BuiltinFunction const* discard = m_dialect.discardFunction({});
+				yulAssert(discard, "");
+				FunctionCall popFunctionCall{
```

Comment:
We also have a helper
```
ExpressionStatement makeDiscardCall(
        langutil::SourceLocation const& _location,
        BuiltinFunction const& _discardFunction,
        Expression&& _expression
)
{
        return {_location, FunctionCall{
                _location,
                Identifier{_location, _discardFunction.name},
                {std::move(_expression)}
        }};
}
```

Perhaps it would make sense moving this into `OptimizerUtils` and using it across places?